adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/DatasourceManager.java

```diff
@@ -98,19 +125,54 @@ public void deleteAllDatasources() {
    */
   @Transactional
   DataImport.MetaData executeImport(Long id, RuntimeParameters runtimeParameters)
-      throws DatasourceNotFoundException, ImporterParameterException, InterpreterParameterException, IOException {
+      throws DatasourceNotFoundException, ImporterParameterException, InterpreterParameterException, IOException {    
     Datasource datasource = getDatasource(id);
-    AdapterConfig adapterConfig = datasource.toAdapterConfig(runtimeParameters);
-    DataImportResponse executionResult = adapter.executeJob(adapterConfig);
-    DataImport dataImport = new DataImport(datasource, executionResult.getData());
+    DataImport dataImport = new DataImport(datasource, "", "FAILED");
+    try {
 
-    DataImport savedDataImport = dataImportRepository.save(dataImport);
+      AdapterConfig adapterConfig = datasource.toAdapterConfig(runtimeParameters);
+      DataImportResponse executionResult = adapter.executeJob(adapterConfig);
+      String responseData = executionResult.getData();
+      dataImport = new DataImport(datasource, responseData);
+
+      if (datasource.getSchema() != null) {
+        validate(datasource, dataImport);
+      }
+
+      DataImport savedDataImport = dataImportRepository.save(dataImport);
 
-    amqpPublisher.publishImportSuccess(id, savedDataImport.getData());
+      amqpPublisher.publishImportSuccess(id, savedDataImport.getData());
 
+      return savedDataImport.getMetaData();
+    } catch (ValidationException e) {
+      System.out.println("exception catch validation");
+      return handleImportWarning(datasource, dataImport, e, runtimeParameters);
```

Comment:
I think this code is not easiliy readable (returning something in a catch clause) => see above how to improve